### PR TITLE
perf: architecture + optimizer optimizations (94.6 min, 4.3% speedup)

### PR DESCRIPTION
**Pull request by vivian-my**
## Summary

- We used [WeCo](https://www.weco.ai/) to optimize nanochat training scripts on d12 (~4.5 min/run), then validated top candidates on d24
- Two rounds of D12 optimization (150 WeCo steps total) + D24 validation
- Best result: **94.6 min** time-to-GPT-2 vs 99 min in Record 6 (**4.3% speedup**)
- Mean CORE = 0.2591 (3 runs), val_bpb = 0.72205

## Run details

**Launch command**:

```
OMP_NUM_THREADS=1 torchrun --standalone --nproc_per_node=8 -m scripts.base_train -- \
    --depth=24 \
    --target-param-data-ratio=8.0 \
    --device-batch-size=16 \
    --fp8 \
    --run=d24_A8s29 \
    --model-tag=d24_A8s29 \
    --sample-every=-1 \
    --save-every=-1 \
    --core-metric-every=999999 \
    --core-metric-max-per-task=-1
```

**Results (3 runs):**

| Metric | Record 6 (baseline) | Ours (mean ± std) | Delta |
|---|---|---|---|
| CORE | 0.2626 | 0.2591 ± 0.0052 | -1.3% (above GPT-2 0.2565) |
| val\_bpb | 0.71816 | 0.72205 ± 0.00004 | +0.54% (slight regression) |
| Training time | 1.65 hr | 1.58 ± 0.001 hr | **-4.3%** |

Individual scores across 3 runs:

| Run | Training time (min) | CORE | val\_bpb |
|---|---|---|---|
| 1 | 94.6 | 0.2540 | 0.72205 |
| 2 | 94.6 | 0.2643 | 0.72208 |
| 3 | 94.6 | 0.2589 | 0.72203 |

**Baseline (Record 6) control** (same hardware, ratio 8.0, unmodified Record 6 code, 1 run):

| Run | Training time (min) | CORE | val\_bpb |
|---|---|---|---|
| 1 | 98.8 | 0.2580 | 0.71827 |

Record 6 training time (98.8 min = 1.65 hr) matches reported in leaderboard, confirming equivalent hardware.

**Per-step speedup breakdown:**
- Architecture changes (smaller windows, fewer VE layers, RoPE 200K): reduces per-step compute
- Removed smear gate: eliminates per-step bigram mixing overhead
- Optimizer buffer pre-allocation: eliminates per-step `torch.stack()` allocations
- Conditional dtype cast in Linear: skips no-op `.to()` calls
- Net per-step improvement: ~1.019s vs ~1.065s baseline (**~4.3% faster per step**)

### Round 1: D12 WeCo Search (150 steps across 3 runs)
- Run A: minimize training time (50 steps)
- Run B: minimize val_bpb (50 steps)
- Run C: multi-token prediction (50 steps, high crash rate)
- Pareto analysis → selected 6 candidates for D24 transfer

### Round 1: D24 Validation (6 candidates + baseline)
- 3/6 candidates passed CORE threshold
- Best: candidate A-8 at 95.3 min, CORE=0.2596

### Round 2: D12 WeCo Refinement (50 steps on A-8)
- Further optimized A-8 → A8-s29
- D12 time: 268.8s → 264.0s, BPB improved (0.8596 → 0.8573)

### Round 2: D24 Validation (3 runs)
- A8-s29: 94.6 min mean, CORE=0.2591 mean, 2/3 runs pass

### `nanochat/gpt.py` — Architecture tuning

All changes reduce per-step compute while preserving model quality. Discovered on d12, validated on d24:

1. **Short window size `seq_len/4` → `seq_len/8`** (512 → 256 tokens): Smaller sliding windows reduce attention FLOPs for short-window layers.
2. **Value embedding frequency: every 2nd → every 3rd layer**: Fewer value embedding layers means fewer parameters and less compute per forward pass.
3. **RoPE base frequency: 100K → 200K**: Slower positional rotation, better long-range generalization.
4. **Smear gate: removed**: Eliminates per-step bigram mixing overhead. Small quality cost, significant speed gain.
5. **Residual lambdas: linear → exponential decay** (1.18→1.06): More principled initialization, better training dynamics.
6. **x0\_lambdas: first-half only** (0.24→0.08→0): Input embedding blending only in early layers where it's most useful. Zero for deep layers.
7. **c\_v init: uniform(-s, s) → uniform(-0.85s, 0.85s)**: Slightly smaller value projection init for stability.
8. **c\_proj init: zeros → uniform(-0.008, 0.008)**: Nonzero output projection init, inspired by modded-nanogpt record #64.
9. **EmbeddingLinear for lm\_head**: Simpler linear class without redundant dtype casting.

### `nanochat/optim.py` — Optimizer efficiency

1. **Pre-allocated buffers**: Replaces per-step `torch.stack()` with `torch.empty()` + in-place copy. Eliminates tensor allocations per optimizer step.
2. **Conditional dtype cast**: Linear class skips `.to()` when weight dtype already matches input. Reduces overhead across all linear layers.

### `scripts/base_train.py` — Hyperparameter tuning

1. **Warmdown ratio: 0.65 → 0.58**: Shorter warmdown, LR stays at peak longer.
2. **Final LR fraction: 0.05 → 0.10**: Higher minimum learning rate prevents optimizer stalling.
3. **Momentum warmup: 0.85→0.97 over 400 steps** (was 300 steps): Longer momentum warmup.
4. **Momentum cooldown: →0.92** (was →0.90): Slightly higher final momentum for stability.

## WeCo D12 Optimisation Results

We ran the search on D12 using WeCo across two rounds:

**Round 1** — 3 parallel optimization tracks (speed, BPB, multi-token prediction), 150 total steps. Pareto analysis selected 6 candidates spanning the speed-quality tradeoff. 

**Round 2** — 50-step refinement on the best D24-validated candidate (A-8). Improved D12 time from 268.8s to 264.0s while recovering BPB. [WeCo Dashboard Round 2](https://dashboard.weco.ai/share/p_5ddFfJSgmJT3MQyXYjBCSFP_JPFbHV)

The best script from D12 Round 2 (step 29) achieves a **5.6% D12 speedup** (264.0s vs 279.6s baseline), which transfers to a **4.3% D24 speedup** (94.6 min vs 98.8 min).

### CODE REVIEW

**Review by svlandeg — COMMENTED** [review]
Why does this PR remove all sorts of inline comments? Does your AI not like them? (anyway, best put them back for easier reviewing)

**Comment by vivian-my** [comment]
> Why does this PR remove all sorts of inline comments? Does your AI not like them? (anyway, best put them back for easier reviewing)

Thanks for the feedback! You're right — the original PR stripped all inline comments during WeCo's automated optimization. I've force-pushed a clean version that preserves all original comments and only shows the actual code changes. Should be much easier to review now.

**Review by svlandeg — COMMENTED** [review]
Ok, I've now done a d14 run on an A100. The run is indeed slightly quicker, but not better in terms of val bpb or CORE. 

Comparing `master` -> this PR:
- tok/sec: 140 -> 144
- wall clock time: 3u23 -> 3u15
- val bpb: 0.838 -> 0.840
- CORE: 0.159 -> 0.150

Image

image

**Comment by svlandeg** [comment]
I might be mistaken, but I think Andrej also wants to see `val_bpb` go down for a leaderboard submission to be included in the table...

**Comment by vivian-my** [comment]
> I might be mistaken, but I think Andrej also wants to see `val_bpb` go down for a leaderboard submission to be included in the table...

Good point. Our val_bpb does regress slightly (+0.004, from 0.71816 to 0.72205). This is the tradeoff: we removed smear, reduced VE coverage, and shrunk sliding windows — all of which reduce per-step compute but slightly hurt model quality. However, the leaderboard rules state: "Your CORE metric must be greater than GPT-2 0.256525. Then you report the total_training_time." val_bpb is reported as a secondary metric, not a hard constraint. Our mean CORE (0.2591) passes the threshold. Happy to hear your preference on whether val_bpb must improve between records.

**Comment by svlandeg** [comment]
I know what it says 🙃 

I also know that for Run 5, Andrej wrote

> ... But the reason I didn't decrease the ratio further (i.e. train shorter) is that while the CORE "safety gap" is large, the val_loss safety gap is smaller - 0.71808, which we want to be below the Run 4 val loss of 0.71854.

I've been trying to clear this up with Andrej whether `val_bpb` should be a hard constraint mentioned in the rules, I'll ping him again about this.
